### PR TITLE
New file for creating synthetic data to add more training data to the tesseract.

### PR DESCRIPTION
**Pull request by JoshuaJAnderson**
# Description

This file will be used to create synthetic data for tesseract to have more data to train on.

[Trello Card](https://trello.com/c/nyFSsSFf/847-as-a-developer-i-want-to-increase-the-accuracy-of-the-tesseract-model-by-adding-new-data-to-be-trained)

[Loom Video](https://www.loom.com/share/c046c5e579144e2fa693af0f2a7d6636)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Change Status

- [ ] Complete, tested, ready to review and merge
- [x] Complete, but not tested (may need new tests)
- [ ] Incomplete/work-in-progress, PR is for discussion/feedback

## Has This Been Tested

- [x] Yes
- [ ] No
- [ ] Not necessary

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] There are no merge conflicts

### CODE REVIEW

**Comment by jraval** [comment]
[As a developer, I want to increase the accuracy of the Tesseract model by finding the best preprocessing steps.](https://trello.com/c/3PndWaDa/848-as-a-developer-i-want-to-increase-the-accuracy-of-the-tesseract-model-by-finding-the-best-preprocessing-steps)

**Review by gianpaolocombatti — CHANGES_REQUESTED** [review]
Hey Josh - Looked at the video and the generator looks awesome, great work! Unfortunately, the data generator in this commit doesn't have any lines of code. Not sure if you uploaded from a different branch. Can you try uploading it again so I can review the actual code? I tried fetching the branch but no luck from that front either.

**Comment by JoshuaJAnderson** [comment]
Not sure why it didn't commit with the code but I tried again and it worked. Thanks for pointing it out!

> Hey Josh - Looked at the video and the generator looks awesome, great work! Unfortunately, the data generator in this commit doesn't have any lines of code. Not sure if you uploaded from a different branch. Can you try uploading it again so I can review the actual code? I tried fetching the branch but no luck from that front either.
Not sure why it was blank, but I recommitted it and it seemed to work. Thanks for letting me know!

**Comment by JoshuaJAnderson** [comment]
There is the whole folder of the characters and I'm not sure if I should commit all of those or not. I felt like it might be too much clutter but if you want to test I can send you the folder.

**Review by nlittlepoole — CHANGES_REQUESTED** [review]
@MicroCapacitor I love what you've got started here but I think what would be most helpful here is to ship this as an experiment so that others can reproduce it. The following is how I would approach doing that

1. create your own experiment folder in `structured_experiments` and name it something like `2022.02.17.synthetic_generation`
2. Copy the code I have in `2022.02.01.test` to your new folder
3. Move `synthetic_data_generator.py` into the folder
4. Modify `synthetic_data_generator.py` so that it generates more than one image (let's say 100) using different strings. Also modify it to save the `.png` and `gt.txt` files to a directory [specified via command line argumet](https://stackoverflow.com/questions/1009860/how-to-read-process-command-line-arguments)
5. Modify the Dockerfile to copy `synthetic_data_generator.py` into the image
6. Modify `experiment.sh` in your folder.  The following should replace the `setup_ground_truth.py` logic. First you want it to create a folder called `synthetic-ground-truth` in `/train/tesstrain/data/`. Then you want to run `synthetic_data_generator.py`, passing it `/train/tesstrain/data/synthetic-ground-truth` as an argument (support added in #4). This will create synthetic images in that folder . Lastly change the `MODEL_NAME` in the make command to use `synthetic` so that Tesseract uses `synthetic-ground-truth` instead of `kaggle-ground-truth`.
7. Run the docker commands in the `experiment.MD` to confirm its working
8. Update the `experiment.MD` with documentation on your experiment

**Comment by derekjpeters** [comment]
[Created function for adding image files to folder with Niger's help.](https://trello.com/c/T3Tn5C6b/866-created-function-for-adding-image-files-to-folder-with-nigers-help)

File: data/synthetic_data_creation/synthetic_data_generator.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+def get_concat_v(im1, im2):
+    dst = Image.new('RGB', (im1.width, im1.height + im2.height))
+    dst.paste(im1, (0, 0))
+    dst.paste(im2, (0, im1.height))
+    return dst
+
+
+# Creates the list of characters from the folder "character_images"
+# These images need to have the letter in the image at the start of the name
+def create_char_list():
```
Comment (bofori20) [inline data/synthetic_data_creation/synthetic_data_generator.py]: Great the script specifies file extensions.

**Review by bofori20 — DISMISSED** [review]
Great work done. We should be able to get in synthetic data to improve the model.

File: structured_experiments/2022.02.17.synthetic_generation/experiment.sh [hunk 2]
```diff
@@ -0,0 +1,5 @@
+echo 'Setting Up Synthetic Test Data for Tesseract Training'
+python synthetic_data_generator.py /train/tesstrain/data/synthetic-ground-truth
```
Comment (nlittlepoole) [inline structured_experiments/2022.02.17.synthetic_generation/experiment.sh:3]: you need to add a line between 1 and 2 and put `mkdir -p /train/tesstrain/data/synthetic-ground-truth` there. If you are running this from a brand new clean clone of the repo then you won't have the `synthetic-ground-truth` folder. You can't commit empty folders to git so this command will make sure the folder exists before running your generations script. The `-p` flag means it won't fail if the folder does happen exist already like it likely does in your `data/` folder.

File: structured_experiments/2022.02.17.synthetic_generation/synthetic_data_generator.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+                                       "woods while she was traveling with her family to Andrea's house which is one "
+                                       "of Mary's friends at school and Mary's muffin was eaten by a bear that sniffed "
+                                       "it from far away. Yes. Mary could be a little weird but that is just how she "
+                                       "is. Back to the real story now. Mary found a fox, took a picture she also "
+                                       "found a rabbit took a picture she found a squirrel and took a picture before "
+                                       "it ran away. But sooner or later she saw two eyes peeking through a bush on "
+                                       "the side of the path she was walking on."))
+"""
+
+for pair in list_of_string_pairs:
```
Comment (nlittlepoole) [inline structured_experiments/2022.02.17.synthetic_generation/synthetic_data_generator.py]: fyi you can do `for file_name, text_data in list_of_string_pairs:`

File: structured_experiments/2022.02.17.synthetic_generation/setup_ground_truth.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import csv
```
Comment (nlittlepoole) [inline structured_experiments/2022.02.17.synthetic_generation/setup_ground_truth.py]: you can delete this file since you aren't using it in your experiment

**Review by nlittlepoole — CHANGES_REQUESTED** [review]
@MicroCapacitor Ok so a few more items:

1. Do you only have one image cut out per image? regardless, you should be adding these to the repository with [git-lfs](https://git-lfs.github.com/). We don't want to pollute the repository with big image files.
2. A lot more fleshed out of an experiment.MD file. This is what future cohorts are going to use to really understand the value of what you have done here. Some questions I'd love to see answered are what is the approach for generating new images and where do you find it in the code? (methodology) What are the theoretical limitations of this approach? What were the training results for tesseract? What can we conclude about using synthetic data from those test results? is it viable? are there limitations? This is the science part of data science that we don't want to lose.

File: structured_experiments/2022.02.17.synthetic_generation/experiment.sh (hunk 2 above)
Comment (JoshuaJAnderson) [reply structured_experiments/2022.02.17.synthetic_generation/experiment.sh:3]: Added such line. Appreciate you typing it out for me!

File: structured_experiments/2022.02.17.synthetic_generation/synthetic_data_generator.py (hunk 3 above)
Comment (JoshuaJAnderson) [reply structured_experiments/2022.02.17.synthetic_generation/synthetic_data_generator.py]: Made such change. I'm pretty sure I've seen it before but forgot about such a thing.

File: structured_experiments/2022.02.17.synthetic_generation/setup_ground_truth.py (hunk 4 above)
Comment (JoshuaJAnderson) [reply structured_experiments/2022.02.17.synthetic_generation/setup_ground_truth.py]: Deleted since yeah, I haven't used it.

**Comment by JoshuaJAnderson** [comment]
Need to figure out if the lfs urls worked on the images I added or not. Else figure out why it's not.
Lambda_storysquadnsynthetic
Ran locally with files in story ground truth. Going to do a few more tests before comparing results.

**Review by nlittlepoole — DISMISSED** [review]
:shipit:

**Review by jinjahninjah — DISMISSED** [review]
Awesome work Joshua! :shipit:

**Review by jinjahninjah — APPROVED** [review]

**Review by nlittlepoole — APPROVED** [review]
👍
